SubsoniciousKit/SubsoniciousKit/Network/Request/NetworkRequest.swift

```diff
@@ -0,0 +1,93 @@
+//
+//  NetworkRequest.swift
+//  SubsoniciousKit
+//
+//  Created by Bilal on 05/07/2020.
+//  Copyright © 2020 Bilal Benlarbi. All rights reserved.
+//
+
+import CryptoKit
+import Foundation
+
+class NetworkRequest {
+
+    let server: Server
+
+    init(server: Server) {
+        self.server = server
+    }
+
+    var path: String { "" }
+
+    var queryItems: [URLQueryItem] {
+        [URLQueryItem(name: ParameterName.version, value: ParameterValue.version),
+         URLQueryItem(name: ParameterName.clientApplication, value: ParameterValue.clientApplication),
+         URLQueryItem(name: ParameterName.dataFormat, value: ParameterValue.dataFormat),
+         URLQueryItem(name: ParameterName.username, value: username),
+         URLQueryItem(name: ParameterName.authenticationToken, value: authenticationToken),
+         URLQueryItem(name: ParameterName.salt, value: salt)]
+    }
+
+    var url: URL? {
+        var urlComponents = URLComponents()
+        urlComponents.scheme = scheme
+        urlComponents.host = host
+        urlComponents.path = path
+        urlComponents.queryItems = queryItems
+
+        return urlComponents.url
+    }
+
+    private lazy var username: String = {
+        server.username
+    }()
+
+    private lazy var authenticationToken: String = {
+        let saltedPassword = server.password + salt
+        let token = try? md5(saltedPassword)
+        return token ?? ""
+    }()
+
+    private lazy var salt: String = {
+        randomString(length: 8)
+    }()
+
+    private lazy var baseURLComponents: [String] = {
+        server.baseURL.components(separatedBy: "://")
+    }()
+
+    private lazy var scheme: String = {
+        baseURLComponents.first ?? ""
+    }()
+
+    private lazy var host: String = {
+        baseURLComponents[safe: 1] ?? ""
+    }()
+}
+
+extension NetworkRequest {
+    enum Error: Swift.Error {
+        case nilURL
+        case nilDataRepresentation
+    }
+}
+
+private extension NetworkRequest {
+
+    typealias ParameterName = Constant.NetworkRequest.ParameterName
+    typealias ParameterValue = Constant.NetworkRequest.ParameterValue
+
+    func randomString(length: Int) -> String {
```

Comment:
I've also thought about that, but extension have an impact on build time and for now I only use this function in that class so I thought it wasn't necessary to use extension :)

Reply:
mmmh... but you already are in an extension, so it is not an argument :P

Reply:
Right, this is part of an extension, but an extension of the `NetworkRequest`, not an extension of a primitive type like String. I prefer to avoid overloading "primitive" types. Of course, if in the future I need a random string, I'll make that func part of a String extension :)

Reply:
okay but you argument for build time impact is no longer relevant. So your argument is you prefer to not overloading primitive types. It's a personal preference so as you wish :)

Reply:
I'll keep it as is for now :)